siva/metadata.go

```diff
@@ -3,154 +3,233 @@ package siva
+func (m *libMetadata) setVersion(v int) {
 	if v != m.CurrentVersion {
 		m.dirty = true
 		m.CurrentVersion = v
 	}
 }
 
-// SetVersion changes the current version.
-func (m *LibraryMetadata) SetID(id string) {
-	if m != nil && id != m.ID {
+// setID changes the current version.
+func (m *libMetadata) setID(id string) {
+	if id != m.ID {
 		m.dirty = true
 		m.ID = id
 	}
 }
 
-// GenerateID creates a new UUID and uses it as ID.
-func (m *LibraryMetadata) GenerateID() error {
+func generateLibID() (string, error) {
 	uuid, err := uuid.NewUUID()
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	m.SetID(uuid.String())
-	return nil
+	return uuid.String(), nil
 }
 
-// Save writes metadata to the library yaml file.
-func (m *LibraryMetadata) Save(fs billy.Filesystem) error {
+func (m *libMetadata) save() error {
+	if !m.dirty {
+		return nil
+	}
+
 	data, err := yaml.Marshal(m)
 	if err != nil {
 		return err
 	}
 
-	tmp := LibraryMetadataFile + ".tmp"
-	defer fs.Remove(tmp)
+	tmp := libraryMetadataFile + ".tmp"
+	defer m.fs.Remove(tmp)
 
-	err = util.WriteFile(fs, tmp, data, 0666)
-	if err != nil {
+	if err = util.WriteFile(m.fs, tmp, data, 0666); err != nil {
 		return err
 	}
 
-	return fs.Rename(tmp, LibraryMetadataFile)
-}
+	if err = m.fs.Rename(tmp, libraryMetadataFile); err != nil {
+		return err
+	}
 
-// Dirty returns true if the metadata was changed and it needs to be written.
-func (m *LibraryMetadata) Dirty() bool {
-	return m.dirty
+	m.dirty = false
+	return nil
 }
 
-// parseLibraryMetadata parses the yaml representation of library metadata.
-func parseLibraryMetadata(d []byte) (*LibraryMetadata, error) {
-	var m LibraryMetadata
-
-	err := yaml.Unmarshal(d, &m)
+func loadLibraryMetadata(fs billy.Filesystem) (*libMetadata, error) {
+	mf, err := fs.Open(libraryMetadataFile)
 	if err != nil {
 		return nil, err
 	}
+	defer mf.Close()
 
-	return &m, nil
-}
+	data, err := ioutil.ReadAll(mf)
+	if err != nil {
+		return nil, err
+	}
 
-// loadLibraryMetadata reads and parses a library metadata file.
-func loadLibraryMetadata(fs billy.Filesystem) (*LibraryMetadata, error) {
-	mf, err := fs.Open(LibraryMetadataFile)
-	if os.IsNotExist(err) {
-		return nil, nil
+	m, err := parseLibraryMetadata(data)
+	if err != nil {
+		return nil, err
 	}
+
+	m.fs = fs
+
+	fi, err := fs.Stat(libraryMetadataFile)
 	if err != nil {
 		return nil, err
 	}
-	defer mf.Close()
 
-	data, err := ioutil.ReadAll(mf)
+	m.modTime = fi.ModTime()
+	m.size = fi.Size()
+
+	return m, nil
+}
+
+func parseLibraryMetadata(d []byte) (*libMetadata, error) {
+	var metadata libMetadata
+
+	err := yaml.Unmarshal(d, &metadata)
 	if err != nil {
 		return nil, err
 	}
 
-	return parseLibraryMetadata(data)
+	return &metadata, nil
 }
 
-// Version describes a bookmark in the siva file.
-type Version struct {
-	// Offset is a position in the siva file.
+// Bookmark represents a valid siva file point to read from.
+type Bookmark struct {
```

Comment:
What's the rationale for changing `Version` to `Bookmark`? It feels more natural that calling `.Version()` returns a `Version` and not a `Bookmark`.